src/backend/sgx/thread.rs

```diff
@@ -160,32 +161,97 @@ fn sgx_enarxcall<'a>(
         }
 
         item::Enarxcall {
-            num: item::enarxcall::Number::RemoveSgxPages,
-            argv: [addr, length, ..],
+            num: item::enarxcall::Number::MmapHost,
+            argv: [addr, length, prot, ..],
             ret,
             ..
         } => {
-            let mut parameters = RemovePages::new(*addr - keep.mem.addr(), *length);
-            ENCLAVE_REMOVE_PAGES
+            // Safety: an `mmap()` call is pointed to a file descriptor of the
+            // created enclave, and can therefore only affect the memory
+            // mappings within the address range given to ENCLAVE_CREATE.
+            let result = unsafe {
+                Map::bytes(*length)
+                    .onto(*addr)
+                    .from(&mut keep.enclave.try_clone().unwrap(), 0)
+                    .with_kind(Shared)
+                    .with(perms::Unknown(*prot as i32))
+            };
+
+            *ret = match result {
+                Ok(map) => {
+                    // Skip `drop()`. The VMA's ownership has been moved to the shim.
+                    std::mem::forget(map);
+                    0
+                }
+                Err(err) => (-err.err.raw_os_error().unwrap()) as usize,
+            };
+
+            Ok(None)
+        }
+
+        item::Enarxcall {
+            num: item::enarxcall::Number::MprotectHost,
+            argv: [addr, length, prot, ..],
+            ret,
+            ..
+        } => {
+            let mem_end = keep.mem.addr() + keep.mem.size();
+            let end = *addr + *length;
+
+            // Check that the span is within the enclave address range:
+            if *addr < keep.mem.addr() || end > mem_end {
+                *ret = libc::EINVAL as usize;
+                return Ok(None);
+            }
+
+            // Safety: the parameters have been verified to be within the shim's
+            // address range.
+            if unsafe { libc::mprotect(*addr as _, *length, *prot as i32) } != 0 {
+                *ret = (-io::Error::last_os_error().raw_os_error().unwrap()) as usize;
+                return Ok(None);
+            }
+
+            // FIXME: Replace `PROT_READ` with `PROT_NONE`, once migrating
+            // to the mainline kernel.  `PROT_READ` is just a workaround for the
+            // patch set version 4.
```

Comment:
Issue link.

Reply:
All of them should point to a single issue.

Reply:
`// TODO: https://github.com/enarx/enarx/issues/1892`